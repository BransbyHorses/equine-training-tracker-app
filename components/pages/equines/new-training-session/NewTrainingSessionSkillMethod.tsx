import React, {useEffect} from "react";
import BackBreadcrumb from "../../../BackBreadcrumb";
import ResponsiveButton from "../../../ResponsiveButton";
import {
	NewSkillTrainingSessionType,
	useNewSkillTrainingSession,
} from "../../../../utils/reducers/trainingSessionReducer";
import { Box, FormControl, InputLabel, Select, MenuItem } from "@mui/material";
import useCollection from "../../../../utils/hooks/useCollection";
import {Skill, TrainingMethod} from "../../../../utils/types";

import PageTitle from "../../../PageTitle";

const NewTrainingSessionSkillMethod = () => {
	const {
		state: { formStage, newTrainingSession },
		dispatch,
	} = useNewSkillTrainingSession();

	const { 
		fetchingData: fetchingSkillsData, 
		collection: skills, 
		error: skillError 
	} = useCollection("skills");

	const { 
		fetchingData: fetchingTrainingMethodData, 
		collection: trainingMethods, 
		error: trainingMethodError 
	} = useCollection("training-methods");

	// useEffect(() => {
	// 	if (!newTrainingSession.skill) {
	// 		dispatch({
	// 			type: NewSkillTrainingSessionType.SET_SKILL,
	// 			payload: "",
	// 		});
	// 	}
	// }, []);

	const changeSkill = (e:any) => {
		const newSkill : Skill = skills.find(skill => e.target.value = skill.id);
		dispatch({
			type: NewSkillTrainingSessionType.SET_SKILL,
			payload: newSkill,
		});
	};

	const changeTrainingMethod = (e:any) => {
		const newTrainingMethod : TrainingMethod = trainingMethods.find(method => e.target.value = method.id);
		dispatch({
			type: NewSkillTrainingSessionType.SET_TRAINING_METHOD,
			payload: newTrainingMethod,
		});
	};


	return (
		<>
			<BackBreadcrumb
                onClick={() => dispatch({ type: NewSkillTrainingSessionType.BACK })}
            />
            <PageTitle title="What skill did you train?" />

			<FormControl fullWidth>
				<InputLabel id="skill-selection">Skill</InputLabel>
				<Select
					value={newTrainingSession.skill?.id || ""}
					name={newTrainingSession.skill?.name}
					label="Skill"
					onChange={changeSkill}
				>
					{skills.map(({ id, name }) => {
						return (
							<MenuItem key={id} value={id}>
								{name}
							</MenuItem>
						);
					})}
				</Select>
			</FormControl>

			<Box sx={{ m: 20 }} />

			<PageTitle title="What method did you use?" />

			<FormControl fullWidth>
				<InputLabel id="method-selection">Training Method</InputLabel>
				<Select
					value={newTrainingSession.trainingMethod?.id || ""}
					name={newTrainingSession.trainingMethod?.name}
					label="Training Method"
					onChange={changeTrainingMethod}
				>
					{trainingMethods.map(({ id, name }) => {
						return (
							<MenuItem key={id} value={id}>
								{name}
							</MenuItem>
						);
					})}
				</Select>
			</FormControl>

			<ResponsiveButton
				disabled={!newTrainingSession.skill || !newTrainingSession.trainingMethod}
				onClick={() => dispatch({ type: NewSkillTrainingSessionType.NEXT })}
			>
				Continue
			</ResponsiveButton>
		</>
	);
};

export default NewTrainingSessionSkillMethod;
