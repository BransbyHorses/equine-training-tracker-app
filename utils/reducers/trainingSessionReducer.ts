import React, { createContext } from "react";
import {
	Skill,
	SkillTrainingSession,
	TrainingEnvironment,
	TrainingMethod,
} from "../types";

enum NewSkillTrainingSessionType {
	NEXT,
	BACK,
	SET_DATE,
	SET_SKILL,
	SET_ENVIRONMENT,
	SET_PROGRESS_CODE,
	SET_TRAINING_METHOD,
	SET_TRAINING_TIME,
	SET_NOTES,
}

interface NewSkillTrainingSessionAction {
	type: NewSkillTrainingSessionType;
	payload: any;
}

const formStages = ["date", "skill_method", "progress", "summary"];

interface NewTrainingSession {
	date: string;
	skill?: Skill;
	environment?: TrainingEnvironment;
	progressCode: string;
	trainingMethod?: TrainingMethod;
	trainingTime?: string;
	notes: "";
}

const newTrainingSession: NewTrainingSession = {
	date: "",
	skill: undefined,
	environment: undefined,
	progressCode: "",
	trainingMethod: undefined,
	trainingTime: undefined,
	notes: "",
};

export const newTrainingSessionState: {
	formStage: string;
	newTrainingSession: NewTrainingSession;
} = {
	formStage: formStages[0],
	newTrainingSession,
};

export function skillTrainingSessionReducer(
	state: { formStage: string; newTrainingSession: NewTrainingSession },
	action: NewSkillTrainingSessionAction
) {
	switch (action.type) {
		case NewSkillTrainingSessionType.NEXT:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) + 1],
			};
		case NewSkillTrainingSessionType.BACK:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) - 1],
			};
		case NewSkillTrainingSessionType.SET_DATE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					date: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_SKILL:
			return { ...state, skill: action.payload };
		case NewSkillTrainingSessionType.SET_ENVIRONMENT:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					environment: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_PROGRESS_CODE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					progressCode: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_METHOD:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingMethod: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_TIME:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingTime: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_NOTES:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					notes: action.payload,
				},
			};
		default:
			return state;
	}
}
