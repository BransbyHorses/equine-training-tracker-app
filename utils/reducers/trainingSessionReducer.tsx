import axios from "axios";
import { Dayjs } from "dayjs";
import React, { Dispatch, useReducer } from "react";
import { Skill, TrainingEnvironment, TrainingMethod } from "../types";

interface NewTrainingSession {
	date?: Dayjs;
	skill?: Skill;
	environment?: TrainingEnvironment;
	progressCode: string;
	trainingMethod?: TrainingMethod;
	trainingTime?: string;
	notes: "";
}

export enum NewSkillTrainingSessionType {
	NEXT,
	BACK,
	GO_TO,
	SET_DATE,
	SET_SKILL,
	SET_ENVIRONMENT,
	SET_PROGRESS_CODE,
	SET_TRAINING_METHOD,
	SET_TRAINING_TIME,
	SET_NOTES,
	SAVE_NEW_TRAINING_SESSION,
}
interface NewSkillTrainingSessionAction {
	type: NewSkillTrainingSessionType;
	payload?: any;
}

export function skillTrainingSessionReducer(
	state: { formStage: string; newTrainingSession: NewTrainingSession },
	action: NewSkillTrainingSessionAction
) {
	switch (action.type) {
		case NewSkillTrainingSessionType.NEXT:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) + 1],
			};
		case NewSkillTrainingSessionType.BACK:
			return {
				...state,
				formStage: formStages[formStages.indexOf(state.formStage) - 1],
			};
		case NewSkillTrainingSessionType.GO_TO:
			return {
				...state,
				formStage: action.payload,
			};
		case NewSkillTrainingSessionType.SET_DATE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					date: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_SKILL:
			return { ...state, skill: action.payload };
		case NewSkillTrainingSessionType.SET_ENVIRONMENT:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					environment: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_PROGRESS_CODE:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					progressCode: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_METHOD:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingMethod: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_TRAINING_TIME:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					trainingTime: action.payload,
				},
			};
		case NewSkillTrainingSessionType.SET_NOTES:
			return {
				...state,
				newTrainingSession: {
					...state.newTrainingSession,
					notes: action.payload,
				},
			};
		default:
			return state;
	}
}

const formStages = [
	"date",
	"skillMethod",
	"environment",
	"progress",
	"summary",
	"success",
];
const newTrainingSession: NewTrainingSession = {
	date: undefined,
	skill: undefined,
	environment: undefined,
	progressCode: "",
	trainingMethod: undefined,
	trainingTime: undefined,
	notes: "",
};

interface NewTrainingSessionState {
	formStage: string;
	newTrainingSession: NewTrainingSession;
}

export const newTrainingSessionInitialState: NewTrainingSessionState = {
	formStage: formStages[0],
	newTrainingSession,
};

interface NewTrainingSessionContextProps {
	state: NewTrainingSessionState;
	dispatch: Dispatch<NewSkillTrainingSessionAction>;
}

export const NewTrainingSessionContext = React.createContext({
	state: newTrainingSessionInitialState,
	dispatch: () => null,
} as NewTrainingSessionContextProps);

export const NewTrainingSessionProvider = (props: any) => {
	const [state, dispatch] = useReducer(
		skillTrainingSessionReducer,
		newTrainingSessionInitialState
	);
	return (
		<NewTrainingSessionContext.Provider value={{ state, dispatch }}>
			{props.children}
		</NewTrainingSessionContext.Provider>
	);
};

export function useNewSkillTrainingSession() {
	const context = React.useContext(NewTrainingSessionContext);
	if (context === undefined) {
		throw new Error(
			"useNewSkillTrainingSession must be used within a NewTrainingSessionProvider"
		);
	}
	return context;
}
