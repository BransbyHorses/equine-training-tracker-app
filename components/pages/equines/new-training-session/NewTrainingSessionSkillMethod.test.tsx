import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import NewTrainingSessionSkillMethod from "./NewTrainingSessionSkillMethod";
import { useNewSkillTrainingSession } from "../../../../utils/reducers/trainingSessionReducer";

jest.mock("../../../../utils/reducers/trainingSessionReducer.tsx", () => ({
	useNewSkillTrainingSession: jest.fn(),
}));

const formStage = "skillMethod";
const newTrainingSession = {
	date: undefined,
	skill: undefined,
	environment: undefined,
	progressCode: "",
	trainingMethod: undefined,
	trainingTime: 0,
	notes: "",
};
const trainingProgramme = {
	trainingCategory: { id: 1, name: "Test Category" },
	equine: { id: 1, name: "Test Equine" },
};

describe("NewTrainingSessionSkillMethod", () => {
	it("renders", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage,
					newTrainingSession,
					trainingProgramme,
				},
			};
		});
		const tree = renderer.create(<NewTrainingSessionSkillMethod />);
		expect(tree).toMatchSnapshot();
	});
	it("disables Continue button when skill and training-method selects are empty", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage,
					newTrainingSession,
					trainingProgramme,
				},
			};
		});
		render(<NewTrainingSessionSkillMethod />);
		expect(screen.getByRole("button", { name: "Continue" })).toBeDisabled();
	});
	it("disables Continue button when skill select is empty", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage,
					newTrainingSession: {
						...newTrainingSession,
						trainingMethod: { id: 1, name: "Test Method" },
					},
					trainingProgramme,
				},
			};
		});
		render(<NewTrainingSessionSkillMethod />);
		expect(screen.getByRole("button", { name: "Continue" })).toBeDisabled();
	});
	it("disabled Continue button when training method select is empty", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage,
					newTrainingSession: {
						...newTrainingSession,
						skill: { id: 1, name: "Test Skill" },
					},
					trainingProgramme,
				},
			};
		});
		render(<NewTrainingSessionSkillMethod />);
		expect(screen.getByRole("button", { name: "Continue" })).toBeDisabled();
	});
	it("allows user to continue to next session", () => {
		(useNewSkillTrainingSession as jest.Mock).mockImplementation(() => {
			return {
				state: {
					formStage,
					newTrainingSession: {
						...newTrainingSession,
						skill: { id: 1, name: "Test Skill" },
						trainingMethod: { id: 1, name: "Test Method" },
					},
					trainingProgramme,
				},
			};
		});
		render(<NewTrainingSessionSkillMethod />);
		expect(screen.getByRole("button", { name: "Continue" })).not.toBeDisabled();
	});
});
