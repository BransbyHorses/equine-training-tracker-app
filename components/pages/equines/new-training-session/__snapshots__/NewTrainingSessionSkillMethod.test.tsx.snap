// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTrainingSessionSkillMethod renders 1`] = `
[
  <nav
    className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-7a84y9-MuiTypography-root-MuiBreadcrumbs-root"
  >
    <ol
      className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
    >
      <li
        className="MuiBreadcrumbs-li"
      >
        <a
          className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-1qt0ygz-MuiTypography-root-MuiLink-root"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m14 7-5 5 5 5V7z"
            />
          </svg>
           Back
        </a>
      </li>
    </ol>
  </nav>,
  <div
    className="MuiBox-root css-17tcgb"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-1lni2c2-MuiTypography-root"
    >
      What skill did you train?
    </h4>
  </div>,
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
    >
      Select a Skill
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl  css-1ualgfl-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-1ftyaf0"
        >
          <span>
            Select a Skill
          </span>
        </legend>
      </fieldset>
    </div>
  </div>,
  <div
    className="MuiBox-root css-uy5y4h"
  />,
  <div
    className="MuiBox-root css-17tcgb"
  >
    <h4
      className="MuiTypography-root MuiTypography-h4 css-1lni2c2-MuiTypography-root"
    >
      What training method did you use?
    </h4>
  </div>,
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
    >
      Selct a Training Method
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl  css-1ualgfl-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value=""
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-1ftyaf0"
        >
          <span>
            Selct a Training Method
          </span>
        </legend>
      </fieldset>
    </div>
  </div>,
  <button
    className="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1g0d0b8-MuiButtonBase-root-MuiButton-root"
    desktopstyles={
      {
        "mt": 3,
        "width": "20%",
      }
    }
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    Continue
  </button>,
]
`;
